Simon-Initiative/oli-torus
# [FEATURE] Paywall guard, payment management, direct payment support via Stripe CMU-191 CMU-190 CMU-193

Represents P4, P5 and P7 of work from Paywall FDD.

Closes #1555
Closes #1554 
Closes #1557 

This work adds in the enforcement of paywalled sections, the ability for a user to redeem a payment code to bypass that the paywall, and an admin only feature to view the issued payment codes and to create and download new batches of codes.  It also adds in the ability for a registered (non-guest) user to make a direct payment, via Stripe. 

There is some infra work here to refine the reusable SortableTableModel and the macro that allows reuse of the common table operation handlers. I refined the extensibility approach to use module attrs. I also had to redo the custom sort approach to make it rely on the more flexible Enum.sort_by/3.

To test this out, do the following:

1. Create a course PRODUCT and require payment and allow a grace period.
2. Create an open and free course section (or launch from LMS and create regular one) and base it on the paid product. Note, to use grace period with "relative to section" strategy you must select a start date for this section.
3. As admin, go to this product and generate some payment codes.
4. Then, in new browser as student attempt to access the O+F course. You should be blocked and redirected to the page prompting you to first create a real account, and then you'll be blocked with the UI that asks you to either make a payment or supply a payment code.
5. Either: Supply one of the payment codes and, voilà, you now have access. OR: Make a direct payment. Use `4242 4242 4242 4242` as a test card number, with any date in the future for the expiration and any security code and zipcode. 
6. Verify you now have access to the course section.
7. From the admin UI, verify the payment record shows up now as being "applied" 

Before we land this to `master` we need to set Tokamak environment variables for the Stripe test key and test secret.
